internal/core/software_renderer/draw_functions.rs

```diff
@@ -656,6 +858,150 @@ pub(super) fn draw_linear_gradient(
+    let fill_col2 = g.flags & 0b100 != 0;
+    let invert_slope = g.flags & 0b1 != 0;
+
+    let y = (line.get() - rect.min_y() + g.top_clip.get()) as i32;
+    let size_y = (rect.height() + g.top_clip.get() + g.bottom_clip.get()) as i32;
+    let start = g.start as i32;
+    let (mut color1, mut color2) = (g.color1, g.color2);
+
+    // Handle vertical gradients (start == 0 case)
+    if g.start == 0 {
+        let p = if invert_slope {
+            (255 - start) * y / size_y
+        } else {
+            start + (255 - start) * y / size_y
+        };
+
+        if (fill_col1 || p >= 0) && (fill_col2 || p < 255) {
+            let gradient_color = interpolate_color(p.clamp(0, 255) as u32, color1, color2);
+
+            // Apply gradient with border radius clipping
+            for (x_offset, pixel) in buffer.iter_mut().enumerate() {
+                let pixel_x = rect.min_x() + extra_left_clip + x_offset as i16;
+                let pixel_y = line.get();
+
+                if is_point_inside_rounded_rect(pixel_x, pixel_y, rect, &rr.radius, rr.width) {
+                    let alpha =
+                        get_rounded_rect_alpha(pixel_x, pixel_y, rect, &rr.radius, rr.width);
+                    if alpha > 0 {
+                        let mut clipped_color = gradient_color;
+                        clipped_color.alpha =
+                            ((clipped_color.alpha as u16 * alpha as u16) / 255) as u8;
+                        pixel.blend(clipped_color);
+                    }
+                }
+            }
+        }
+        return;
+    }
+
+    // Handle diagonal gradients with border radius clipping
+    let size_x = (rect.width() + g.left_clip.get() + g.right_clip.get()) as i32;
+
+    if !invert_slope {
+        core::mem::swap(&mut color1, &mut color2);
+    }
+
+    let dr = (((color2.red as i32 - color1.red as i32) * start) << 15) / (255 * size_x);
+    let dg = (((color2.green as i32 - color1.green as i32) * start) << 15) / (255 * size_x);
+    let db = (((color2.blue as i32 - color1.blue as i32) * start) << 15) / (255 * size_x);
+    let da = (((color2.alpha as i32 - color1.alpha as i32) * start) << 15) / (255 * size_x);
+
+    // Calculate initial gradient position for this line
+    let gradient_x = if invert_slope {
+        (y * size_x * (255 - start)) / (size_y * start)
+    } else {
+        (size_y - y) * size_x * (255 - start) / (size_y * start)
+    } + g.left_clip.get() as i32
+        + extra_left_clip as i32;
+
+    // Process each pixel with border radius clipping
+    for (x_offset, pixel) in buffer.iter_mut().enumerate() {
+        let pixel_x = rect.min_x() + extra_left_clip + x_offset as i16;
+        let pixel_y = line.get();
+
+        // Check if pixel is inside the rounded rectangle
+        if is_point_inside_rounded_rect(pixel_x, pixel_y, rect, &rr.radius, rr.width) {
+            let alpha = get_rounded_rect_alpha(pixel_x, pixel_y, rect, &rr.radius, rr.width);
```

Comment:
Is that not quite slow to do for every pixel?
There should ideally be some caching and compute the region which can be rendered without having to call this.